### ISSUE TITLE
BadCertificateRevocationUnknown - Unable to connect with using security - OPC Publisher

### ISSUE THREAD

**Issue by JohannesM24** [Azure/Industrial-IoT#1771]
**Describe the bug**
When trying to connect to an OPC UA server with using security, we get the following error:

```
2022-09-08 09:34:55.746 +02:00 [DBG] Failed to connect to opc.tcp://xxx:4814/xxx: One or more errors occurred. (BadCertificateRevocationUnknown) - try again...
2022-09-08 09:34:55.746 +02:00 [WRN] Failed to create session 'opc.tcp://xxx:4814/xxx_467F59DB' due to One or more errors occurred. (One or more errors occurred. (BadCertificateRevocationUnknown))
```

**Expected behavior**
The connection should also build up with using security (as it is working without security options). 

**Desktop:**
 - Windows Server 2019
 - OPC Publisher 2.8.3
 - stand alone mode

**Additional context**
We have checked the certificates and everything looks good (being trusted on client side). We have the same setup working in our productive system that runs on publisher 2.8.1 and now we wanted to also enable security for the development environment.

It also did not work in 2.8.2 in the dev environment.

What are the steps we could follow to verify our setups / debug the origin of BadCertificateRevocationUnknown?
Have there been any change to the certificates from 2.8.1 to 2.8.3 ?

**Comment by monikavar** [comment]
@JohannesM24  could you please provide the support bundle to investigate this further?

**Comment by mregen** [comment]
Hi @JohannesM24, the error means there is a revocation list missing for a CA certificate on the client side, e.g. if a CA is used in a trusted folder the CRL should be added too. The CRL is hosted in a e.g. `trusted/crl` when the CA cert is under `trusted/certs`.

The default may have been revised since version 2.8.1 in an effort to connect secure by design. Now the default is to reject unknown revocation status.

@monikavar please check if the publisher needs an option to suppress the error by setting the application configuration in SecurityConfigEx.cs:
```c#
            var options = applicationConfigurationBuilder
                .AddSecurityConfiguration(
                    securityConfig.ApplicationCertificate.SubjectName.Replace("localhost", hostname),
                    securityConfig.PkiRootPath)
        --> add setting        .SetRejectUnknownRevocationStatus(false or flag)
                .SetAutoAcceptUntrustedCertificates(securityConfig.AutoAcceptUntrustedCertificates)
```

 `ISecurityConfig` should have a flag to set `RejectUnknownRevocationStatus`

**Comment by JohannesM24** [comment]
Hi @mregen ,
thanks for your answer. We have now added the revocation lists for our intermediate and ca cert in the folder structure you suggested. It also seems to get found, as we can see `2022-10-18 16:54:02.867 +02:00 [INF] Trusted peer store has 2 CRLs.`
But even before that already the `BadCertificateRevocationUnkown ` Error happens.

Please find here the longer logs:


```
2022-10-18 16:54:02.503 +02:00 [INF] Module Host started.
2022-10-18 16:54:02.513 +02:00 [INF] Creating new worker with id StandalonePublisher_0
2022-10-18 16:54:02.514 +02:00 [INF] Starting worker 'StandalonePublisher_0'...
2022-10-18 16:54:02.517 +02:00 [INF] Starting worker StandalonePublisher_0: {}
2022-10-18 16:54:02.517 +02:00 [DBG] Worker StandalonePublisher_0 sending heartbeat...
2022-10-18 16:54:02.524 +02:00 [DBG] Worker update with "Keep" instruction for job null.
2022-10-18 16:54:02.526 +02:00 [DBG] Worker StandalonePublisher_0 starting...
2022-10-18 16:54:02.526 +02:00 [DBG] Worker StandalonePublisher_0 tries querying available job...
2022-10-18 16:54:02.530 +02:00 [DBG] Process worker StandalonePublisher_0, with job opc.tcp://opc_server_redacted_B03893F4.
2022-10-18 16:54:02.535 +02:00 [INF] Worker StandalonePublisher_0 processing job opc.tcp://opc_server_redacted_B03893F4, mode: "Active"
2022-10-18 16:54:02.676 +02:00 [INF] Imported the private key for [thumbprint_leaf_cert_redacted].
2022-10-18 16:54:02.688 +02:00 [INF] Own certificate Subject 'leaf_cert_redacted' (thumbprint: thumbprint_leaf_cert_redacted) loaded.
2022-10-18 16:54:02.691 +02:00 [INF] Checking application instance certificate.
2022-10-18 16:54:02.711 +02:00 [INF] Imported the private key for [thumbprint_leaf_cert_redacted].
2022-10-18 16:54:02.712 +02:00 [INF] Check certificate: [leaf_cert_redacted] [thumbprint_leaf_cert_redacted]
2022-10-18 16:54:02.714 +02:00 [INF] Check application instance certificate. [leaf_cert_redacted] [thumbprint_leaf_cert_redacted]
2022-10-18 16:54:02.834 +02:00 [WRN] Certificate Validation failed. Reason=BadCertificateRevocationUnknown. [leaf_cert_redacted] [thumbprint_leaf_cert_redacted]
2022-10-18 16:54:02.835 +02:00 [INF] Rejecting peer certificate thumbprint_leaf_cert_redacted, 'leaf_cert_redacted' because of BadCertificateRevocationUnknown.
2022-10-18 16:54:02.836 +02:00 [WRN] Application Certificate Validation suppressed BadCertificateRevocationUnknown
2022-10-18 16:54:02.836 +02:00 [WRN] Validation errors suppressed:  [leaf_cert_redacted] [thumbprint_leaf_cert_redacted]
2022-10-18 16:54:02.844 +02:00 [INF] Updated the ApplicationUri: urn:urn_redacted: --> urn:thg-iotedged01:THG-IOTEDGED01:microsoft:
2022-10-18 16:54:02.844 +02:00 [INF] Using the ApplicationUri: urn:thg-iotedged01:THG-IOTEDGED01:microsoft:
2022-10-18 16:54:02.861 +02:00 [INF] Application own certificate store contains 1 certs.
2022-10-18 16:54:02.861 +02:00 [INF] 01: Subject 'leaf_cert_redacted' (thumbprint: thumbprint_leaf_cert_redacted)
2022-10-18 16:54:02.862 +02:00 [INF] Trusted issuer store contains 0 certs.
2022-10-18 16:54:02.863 +02:00 [INF] Trusted issuer store has 0 CRLs.
2022-10-18 16:54:02.865 +02:00 [INF] Trusted peer store contains 3 certs.
2022-10-18 16:54:02.866 +02:00 [INF] 01: Subject 'leaf_cert_redacted' (thumbprint: thumbprint_leaf_cert_redacted)
2022-10-18 16:54:02.866 +02:00 [INF] 02: Subject 'intermediate_cert_redacted' (thumbprint: intermediate_cert_thumbprint_redacted)
2022-10-18 16:54:02.866 +02:00 [INF] 03: Subject 'ca_cert_redacted' (thumbprint: ca_cert_thumbprint_redacted)
2022-10-18 16:54:02.867 +02:00 [INF] Trusted peer store has 2 CRLs.
2022-10-18 16:54:02.879 +02:00 [INF] 01: Issuer 'intermediate_cert_redacted', Next update time '"2022-10-13T19:57:22.0000000Z"'
2022-10-18 16:54:02.880 +02:00 [INF] 02: Issuer 'ca_cert_redacted', Next update time '"2023-01-05T20:36:26.0000000Z"'
2022-10-18 16:54:02.881 +02:00 [INF] Rejected certificate store contains 0 certs.
2022-10-18 16:54:02.897 +02:00 [INF] Subscription '100' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.898 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.904 +02:00 [INF] Subscription '100' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.904 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.904 +02:00 [INF] Subscription '500' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.904 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.905 +02:00 [INF] Subscription '500' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.905 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.905 +02:00 [INF] Subscription '10' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.905 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.905 +02:00 [INF] Subscription '10' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.905 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.905 +02:00 [INF] Subscription '1000' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.905 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.906 +02:00 [INF] Subscription '1000' registered/updated in session 'opc.tcp://opc_server_redacted_B03893F4' in state "Init"
2022-10-18 16:54:02.906 +02:00 [DBG] Runner Keepalive reset due to  triggerKeepAlive
2022-10-18 16:54:02.907 +02:00 [DBG] Initializing session 'opc.tcp://opc_server_redacted_B03893F4'...
2022-10-18 16:54:02.927 +02:00 [INF] Job opc.tcp://opc_server_redacted_B03893F4 started.
2022-10-18 16:54:03.068 +02:00 [INF] ChannelId 0: in "Connecting" state.
2022-10-18 16:54:03.089 +02:00 [INF] ChannelId 0: in "Opening" state.
2022-10-18 16:54:03.090 +02:00 [INF] ChannelId 0: Token #0 created. CreatedAt=14:54:03.090. Lifetime=3600000.
2022-10-18 16:54:03.111 +02:00 [INF] Security Policy: http://opcfoundation.org/UA/SecurityPolicy#None
2022-10-18 16:54:03.111 +02:00 [INF] Sender Certificate: (none)
2022-10-18 16:54:03.120 +02:00 [INF] ChannelId 1992: Token #1 activated. CreatedAt=14:54:03.090. Lifetime=3600000.
2022-10-18 16:54:03.120 +02:00 [INF] ChannelId 1992: in "Open" state.
2022-10-18 16:54:03.121 +02:00 [INF] ChannelId 1992: Token Expiry "2022-10-18T15:54:03.0906138Z", renewal scheduled in 2699976 ms.
2022-10-18 16:54:03.260 +02:00 [INF] ChannelId 1992: in "Closing" state.
2022-10-18 16:54:03.265 +02:00 [WRN] ChannelId 1992: Force reconnect reason=BadConnectionClosed 'Remote side closed connection'
2022-10-18 16:54:03.266 +02:00 [INF] ChannelId 1992: in "Closed" state.
2022-10-18 16:54:03.267 +02:00 [INF] ChannelId 1992: CLIENTCHANNEL SOCKET CLOSED: 00FC006A
2022-10-18 16:54:03.269 +02:00 [INF] Creating session 'opc.tcp://opc_server_redacted_B03893F4' for endpoint 'opc.tcp://opc_server_redacted'...
2022-10-18 16:54:03.274 +02:00 [INF] ChannelId 0: in "Connecting" state.
2022-10-18 16:54:03.277 +02:00 [INF] ChannelId 0: in "Opening" state.
2022-10-18 16:54:03.278 +02:00 [INF] ChannelId 0: Token #0 created. CreatedAt=14:54:03.278. Lifetime=3600000.
2022-10-18 16:54:03.280 +02:00 [INF] Security Policy: http://opcfoundation.org/UA/SecurityPolicy#None
2022-10-18 16:54:03.280 +02:00 [INF] Sender Certificate: (none)
2022-10-18 16:54:03.280 +02:00 [INF] ChannelId 1993: Token #1 activated. CreatedAt=14:54:03.278. Lifetime=3600000.
2022-10-18 16:54:03.280 +02:00 [INF] ChannelId 1993: in "Open" state.
2022-10-18 16:54:03.280 +02:00 [INF] ChannelId 1993: Token Expiry "2022-10-18T15:54:03.2780618Z", renewal scheduled in 2699997 ms.
2022-10-18 16:54:03.289 +02:00 [INF] ChannelId 1993: in "Closing" state.
2022-10-18 16:54:03.290 +02:00 [WRN] ChannelId 1993: Force reconnect reason=BadConnectionClosed 'Remote side closed connection'
2022-10-18 16:54:03.290 +02:00 [INF] ChannelId 1993: in "Closed" state.
2022-10-18 16:54:03.290 +02:00 [INF] ChannelId 1993: CLIENTCHANNEL SOCKET CLOSED: 002C160B
2022-10-18 16:54:03.317 +02:00 [INF] Azure IIoT opc.tcp://opc_server_redacted_B03893F4 took "00:00:00.0477872".
2022-10-18 16:54:03.317 +02:00 [DBG] Failed to connect to opc.tcp://opc_server_redacted: One or more errors occurred. (BadCertificateRevocationUnknown) - try again...
2022-10-18 16:54:03.318 +02:00 [WRN] Failed to create session 'opc.tcp://opc_server_redacted_B03893F4' due to One or more errors occurred. (One or more errors occurred. (BadCertificateRevocationUnknown))
```

**Comment by bqstony** [comment]
> Hi @JohannesM24, the error means there is a revocation list missing for a CA certificate on the client side, e.g. if a CA is used in a trusted folder the CRL should be added too. The CRL is hosted in a e.g. `trusted/crl` when the CA cert is under `trusted/certs`.
> 
> The default may have been revised since version 2.8.1 in an effort to connect secure by design. Now the default is to reject unknown revocation status.
> 
> @monikavar please check if the publisher needs an option to suppress the error by setting the application configuration in SecurityConfigEx.cs:
> 
> ```cs
>             var options = applicationConfigurationBuilder
>                 .AddSecurityConfiguration(
>                     securityConfig.ApplicationCertificate.SubjectName.Replace("localhost", hostname),
>                     securityConfig.PkiRootPath)
>         --> add setting        .SetRejectUnknownRevocationStatus(false or flag)
>                 .SetAutoAcceptUntrustedCertificates(securityConfig.AutoAcceptUntrustedCertificates)
> ```
> 
> `ISecurityConfig` should have a flag to set `RejectUnknownRevocationStatus`

I have the same issue. Can you please descripe how to setup the CRL in the pki folder structure.

my own certificatin has the entry "CRL Distribution Point", i hope this is enough.

**Comment by bqstony** [comment]
I Think i have found it unter: https://reference.opcfoundation.org/v104/GDS/docs/F.1/ and here https://github.com/OPCFoundation/UA-.NETStandard/blob/master/Docs/Certificates.md

**Comment by mregen** [comment]
Hi @JohannesM24, one of the CRL is already expired at the time of connection:
```
01: Issuer 'intermediate_cert_redacted', Next update time '"2022-10-13T19:57:22.0000000Z"'
```

**Comment by marcschier** [comment]
As change was made, I will close assuming this is addressed in 2.8.4 and above.  If not, please re-open, and then apologies in advance, I was misreading the state of the issue.
